config/config.go

```diff
@@ -414,12 +418,23 @@ func parseHcl(parser *hclparse.Parser, hcl string, filename string) (file *hcl.F
 		}
 	}()
 
-	file, parseDiagnostics := parser.ParseHCL([]byte(hcl), filename)
-	if parseDiagnostics != nil && parseDiagnostics.HasErrors() {
-		return nil, parseDiagnostics
+	if filepath.Ext(filename) == ".hcl" {
```

Comment:
Here are a few other options:

### Option 1: Trial and error

The code can try to parse as HCL first, and then if there are errors, fall back to parsing as JSON (or other way around). Since most of the implementations out there will be HCL, I think doing HCL first makes sense for efficiency reasons.

Note that for this implementation, you need to create a `parseJson` function as opposed to doing the if check here, because you won't be able to handle the panics correctly otherwise. The idea would be to have three functions:

```
parseHclOrJson
parseHcl
parseJson
```

where `parseHcl` remains the same as it was before.

A problem with this approach is that you don't want to report the json parsing errors if the file was meant to be parsed as HCL (or vice versa), so you will need to make sure to report both parsing errors.


### Option 2: Heuristics based approach

Since HCL compatible JSON syntax is always an object, you can check if the first non-whitespace character is `{` and the last non-whitespace character is `}`. This could be brittle if the terragrunt syntax changes and the top level item is a literal map expression in hcl2.

### Option 3: Command line option

This could be configured as a command line option `--terragrunt-syntax-type`, so that it is explicit. This would be a cumbersome interface though.

---

Having laid out these options, I'd vote for keeping the file extension approach you have, with a minor adjustment. We should treat `json` as the special case, and always default to `hcl`. So the if condition should be:

```
if filepath.Ext(filename) == ".json" {
    // parse as json
}
// parse as HCL
```

This way, we preserve the old behavior for any file extension except for those that have `.json`, and I can't imagine anyone intending to use terragrunt with a `.json` syntax expecting HCL parsing (vs you might expect that with something like a `.terragrunt` file).